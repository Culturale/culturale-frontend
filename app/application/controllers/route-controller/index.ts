export * from './route-controller';
export * from './route-controller.interface';
