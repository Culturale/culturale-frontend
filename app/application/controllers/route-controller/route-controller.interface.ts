import type { Controller } from '~/application/controllers/controller.interface';
import type { IRoute } from '~/domain';

/**
 * Module that controls the route assigned to the driver
 *
 */
export interface IRouteController extends Controller {
  route: IRoute;

  setup: () => Promise<void>;

  setRoute: (route: IRoute) => void;
}
