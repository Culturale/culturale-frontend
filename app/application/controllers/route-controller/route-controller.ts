import AsyncStorage from '@react-native-community/async-storage';
import { action, makeObservable, observable } from 'mobx';
import { makePersistable } from 'mobx-persist-store';

import type { IRoute} from '~/domain';
import { Route } from '~/domain';
import type { IInfrastructure } from '~/infrastructure';
import type { RequestSubjectsMap } from '~/observables';
import { RequestSubject } from '~/observables';

import type { IRouteController } from './route-controller.interface';

export enum RouteRequests {
  RequestRouteDocumentID = 'REQUEST_ROUTE_DOCUMENT_ID',
}

export class RouteController implements IRouteController {
  public route: IRoute;
  public driverId: string;
  public routeId: string;
  // @ts-ignore
  private infrastructure: IInfrastructure;

  public requestsSubjectsMap: RequestSubjectsMap<RouteRequests> = {
    [RouteRequests.RequestRouteDocumentID]: new RequestSubject<string>(
      RouteRequests.RequestRouteDocumentID
    )
  };

  constructor(infrastructure?: IInfrastructure) {
    this.route = new Route();
    this.infrastructure = infrastructure;
    makeObservable(this, {
      driverId: observable,
      route: observable,
      setRoute: action
    });
  }

  public setRoute(route: IRoute) {
    this.route = route;
  }

  setDriverId: (driverId: string) => void;

  public async setup(): Promise<void> {
    await makePersistable(
      this,
      {
        name: 'RouteController',
        properties: ['driverId'],
        storage: AsyncStorage
      },
      { fireImmediately: true }
    );
  }
}
