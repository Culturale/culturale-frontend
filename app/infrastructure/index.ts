export * from './infrastructure.interface';
export * from './infrastructure';
export * from './services';
export * from './infrastructure.mock';
export * from './api';
