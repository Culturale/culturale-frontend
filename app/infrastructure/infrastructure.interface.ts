import type { IAPI } from './api';
import type {
  ILocationService
} from './services';

export interface IInfrastructure {
  services: {
    location: ILocationService;
  };
  api: IAPI;
}
