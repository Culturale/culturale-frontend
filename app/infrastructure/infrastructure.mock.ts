import { MockAPI } from './api';
import type { IInfrastructure } from './infrastructure.interface';
import type {
  IMockLocationService} from './services';
import { MockLocationService
} from './services';

export interface IMockInfrastructure extends IInfrastructure {
  readonly services: {
    location: IMockLocationService;
  };

  readonly api: MockAPI;
}

export class MockInfrastructure implements IMockInfrastructure {
  public readonly services: {
    location: IMockLocationService;
  };

  public readonly api: MockAPI;

  constructor() {
    this.services = {
      location: new MockLocationService()
    };
    this.api = new MockAPI('test-url');
  }
}
