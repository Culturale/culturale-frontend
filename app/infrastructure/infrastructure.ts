import env from '~/config/env';

import type { IAPI } from './api';
import { API } from './api';
import type { IInfrastructure } from './infrastructure.interface';
import type { ILocationService} from './services';
import { LocationService } from './services';

export class Infrastructure implements IInfrastructure {
  public readonly services: {
    location: ILocationService;
  };

  public readonly api: IAPI;

  constructor() {
    this.services = {
      location: new LocationService()
    };
    this.api = new API(env.API_HOST);
  }
}
