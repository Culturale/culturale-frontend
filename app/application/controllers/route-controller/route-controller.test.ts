import type { IInfrastructure} from '~/infrastructure';
import { MockInfrastructure } from '~/infrastructure';

import { RouteController } from './route-controller';
import type { IRouteController } from './route-controller.interface';

describe('RouteController', function () {
  let instance: IRouteController;
  const infrastructure: IInfrastructure = new MockInfrastructure();

  beforeEach(function () {
    instance = new RouteController(infrastructure);
  });

  it('can be created', function () {
    expect(instance).toMatchSnapshot();
  });
});
