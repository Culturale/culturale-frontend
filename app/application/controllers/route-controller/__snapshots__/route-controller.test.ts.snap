// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouteController can be created 1`] = `
RouteController {
  "driverId": undefined,
  "infrastructure": MockInfrastructure {
    "api": MockAPI {
      "baseURL": "test-url",
      "error": undefined,
      "failedTasks": Object {},
      "returnAPIError": false,
      "token": undefined,
    },
    "services": Object {
      "location": MockLocationService {
        "gpsEnabled": true,
        "location": Geocode {
          "latitude": 4,
          "longitude": 3,
        },
        "locationUpdatesSubscription": undefined,
        "setupCalled": false,
      },
    },
  },
  "requestsSubjectsMap": Object {
    "REQUEST_ROUTE_DOCUMENT_ID": RequestSubject {
      "_value": "IDLE",
      "closed": false,
      "currentObservers": null,
      "hasCompleted": false,
      "hasError": false,
      "hasFailed": false,
      "isStopped": false,
      "observers": Array [],
      "requestID": "REQUEST_ROUTE_DOCUMENT_ID",
      "requestStatus": "IDLE",
      "thrownError": null,
    },
  },
  "route": Route {
    "route": Array [],
    "routeState": "FULLY_SEQUENCED",
  },
  "routeId": undefined,
  "setDriverId": undefined,
}
`;
